### PR TITLE
M408

### PR DESCRIPTION
**Pull request by akaJes**
http://reprap.org/wiki/G-code#M408:_Report_JSON-style_response
+802 bytes

### CODE REVIEW

**Comment by bgort** [comment]
Out of curiosity, what's the (or your) use case for this?

File: .gitignore [hunk 1]
```diff
@@ -136,3 +136,7 @@ CMakeListsPrivate.txt
 
 #CLion
 cmake-build-*
+.pioenvs
```
Comment (bgort) [inline .gitignore]: This and the following one are already covered by line 120.

Comment (akaJes) [reply .gitignore]: these were auto appended with gui

**Comment by akaJes** [comment]
@bgort I would like to create a web interface
@thinkyhead have you some seconds to look at this PR ?

**Comment by bgort** [comment]
I don't see any reason not to merge this.  It's conditionally compiled, and as long as it works, I can't see how it hurts anything.  Let's give @thinkyhead a bit longer to comment and then I'll merge it.

File: .travis.yml [hunk 2]
```diff
@@ -258,6 +258,12 @@ script:
   - opt_enable PRINTCOUNTER
   - build_marlin
   #
+  # Test REPORT_M408_JSON
+  #
+  - restore_configs
+  - opt_enable REPORT_M408_JSON
+  - build_marlin
+  #
```
Comment (thinkyhead) [inline .travis.yml]: Enable the option in one of the other tests, rather than requiring an additional build.

File: Marlin/Marlin_main.cpp [hunk 3]
```diff
@@ -8869,6 +8869,93 @@ inline void gcode_M400() { stepper.synchronize(); }
...
+    LOOP_XYZ(i) {
+      if ( i != X_AXIS )
+        SERIAL_PROTOCOLCHAR(',');
+      SERIAL_PROTOCOLCHAR(axis_homed[i]?'1':'0');
+    }
+    #if ENABLED(SDSUPPORT)
+      SERIAL_PROTOCOLPGM("],\"fraction_printed\": ");
+      SERIAL_PROTOCOL_F(card.percentDone() * 0.01, 3);
+    #else
+      SERIAL_PROTOCOLPGM("]");
```
Comment (thinkyhead) [inline Marlin/Marlin_main.cpp]:
```cpp
SERIAL_PROTOCOLCHAR(']');
```

File: Marlin/Marlin_main.cpp [hunk 4]
```diff
@@ -8869,6 +8869,93 @@ inline void gcode_M400() { stepper.synchronize(); }
...
+      SERIAL_PROTOCOL(-1);
+    #endif
+    #if HAS_TEMP_HOTEND
+      HOTEND_LOOP() {
+        SERIAL_PROTOCOLCHAR(',');
+        SERIAL_PROTOCOL(thermalManager.degTargetHotend(e));
+      }
+    #endif
+    //skipped standbay & hstat
+    SERIAL_PROTOCOLPGM("],\"pos\": [");
```
Comment (thinkyhead) [inline Marlin/Marlin_main.cpp:8910]: These custom print statements that create key/value strings should probably be implemented as a small function taking the key and value. The code will likely end up smaller.

File: Marlin/Marlin_main.cpp [hunk 5]
```diff
@@ -8869,6 +8869,93 @@ inline void gcode_M400() { stepper.synchronize(); }
...
+  /**
+   * M408: Report JSON-style response to serial
+   */
+  inline void gcode_M408() {
+    // I=idle, P=printing from SD card, S=stopped (i.e. needs a reset), C=running config file (i.e starting up), A=paused, D=pausing, R=resuming from a pause, B=busy (e.g. running a macro),
+    SERIAL_PROTOCOLPGM("{\"status\": \"");
+    if (!planner.blocks_queued())
+      SERIAL_PROTOCOLCHAR('I'); // IDLING
+    else
+    if (IS_SD_PRINTING || print_job_timer.isRunning())
```
Comment (thinkyhead) [inline Marlin/Marlin_main.cpp]: Please put `else if` together on the same line.

File: Marlin/Marlin_main.cpp [hunk 6]
```diff
@@ -8869,6 +8869,93 @@ inline void gcode_M400() { stepper.synchronize(); }
...
+      SERIAL_PROTOCOLCHAR('P'); // SD PRINTING
+    else
+    if (print_job_timer.isPaused())
+      SERIAL_PROTOCOLCHAR('A'); // PAUSED
+    else
+      SERIAL_PROTOCOLCHAR('B'); // SOMETHING ELSE
```
Comment (thinkyhead) [inline Marlin/Marlin_main.cpp]: Please put `else if` keywords together on the same line. Or use this:
```cpp
SERIAL_PROTOCOLCHAR(
  !planner.blocks_queued()                        ? 'I' : // IDLING
  (IS_SD_PRINTING || print_job_timer.isRunning()) ? 'P' : // SD Printing
  print_job_timer.isPaused()                      ? 'A' : // PAUSED
                                                    'B'   // SOMETHING ELSE
);
```

File: Marlin/Marlin_main.cpp [hunk 7]
```diff
@@ -8869,6 +8869,93 @@ inline void gcode_M400() { stepper.synchronize(); }
...
+        SERIAL_PROTOCOLCHAR(',');
+      SERIAL_PROTOCOLCHAR(axis_homed[i]?'1':'0');
+    }
+    #if ENABLED(SDSUPPORT)
+      SERIAL_PROTOCOLPGM("],\"fraction_printed\": ");
+      SERIAL_PROTOCOL_F(card.percentDone() * 0.01, 3);
+    #else
+      SERIAL_PROTOCOLPGM("]");
+    #endif
+    SERIAL_PROTOCOLLNPGM("}");
```
Comment (thinkyhead) [inline Marlin/Marlin_main.cpp]:
```cpp
SERIAL_PROTOCOLCHAR('}');
```

File: .travis.yml (hunk 2 above)
Comment (akaJes) [reply .travis.yml]: which is more prefered? PRINTCOUNTER ?

File: Marlin/Marlin_main.cpp (hunk 4 above)
Comment (akaJes) [reply Marlin/Marlin_main.cpp:8910]: I thought about this, but don't see good solution yet

**Comment by thinkyhead** [comment]
Something screwy is going on here. The base was changed to `bugfix-2.0.x` on July 31, but the PR has changes to `Marlin_main.cpp`, which is only in the 1.1 layout.

Please redo.

File: Marlin/Marlin_main.cpp (hunk 4 above)
Comment (thinkyhead) [reply Marlin/Marlin_main.cpp:8910]: A single key-value output would be…
```cpp
void floatkey(const char *pstr, const float &val, const bool comma=false) {
  SERIAL_CHAR('"'); serialprintPGM(pstr); SERIAL_CHAR('"');
  SERIAL_CHAR(':'); SERIAL_ECHO(val);
  if (comma) SERIAL_ECHO(',');
}
```
And a function for XYZ elements…
```cpp
void xyzkey(const char *pstr, const float val[XYZ], const bool comma=false) {
  SERIAL_CHAR('"'); serialprintPGM(pstr); SERIAL_CHAR('"');
  SERIAL_ECHO(":[");
  SERIAL_ECHO(val[X_AXIS]); SERIAL_CHAR(',');
  SERIAL_ECHO(val[Y_AXIS]); SERIAL_CHAR(',');
  SERIAL_ECHO(val[Z_AXIS]);
  SERIAL_CHAR(']');
  if (comma) SERIAL_ECHO(',');
}

**Pull request closed without merging**